### PR TITLE
feat: Add a function to compute closest points

### PR DESCRIPTION
**Pull request by rhit-chois3**
# Description

This PR adds a `closestPoint` function which takes in a shape and a point, and returns the closest point on that shape to the given point.

# Implementation strategy and design decisions

This PR changes the type of the `clamp` function to accept `ad.Num`s for all parameters, instead of requiring the bounds of the range to be fixed `number`s. The body of the function is the same.

# Examples with steps to reproduce them

The example in `packages/examples/src/closest-point/` plots the closest point to several basic shapes for a single query point (resampling picks new shapes and a new point):

image

# Checklist

- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new ESLint warnings
- [X] I have reviewed any generated changes to the `diagrams/` folder

### CODE REVIEW

File: packages/core/src/contrib/ClosestPoint.test.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import { makeCircle } from "shapes/Circle";
+import { makeRectangle } from "shapes/Rectangle";
+import { Context, makeCanvas, simpleContext } from "shapes/Samplers";
+import { Shape } from "types/shapes";
+import { black, floatV, vectorV } from "utils/Util";
+import { compDict } from "./Functions";
+
+const canvas = makeCanvas(800, 700);
+
+const compareClosestPoint = (
```
Comment (cmumatt) [inline packages/core/src/contrib/ClosestPoint.test.ts:11]: May be helpful to add a /** */ comment to each function to document its meaning and the meaning of its input / output values.

File: packages/core/src/contrib/ClosestPoint.test.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  //
```
Comment (cmumatt) [inline packages/core/src/contrib/ClosestPoint.test.ts:12]:
```suggestion
```

If this line is not needed here, maybe just delete it?

File: packages/core/src/contrib/ClosestPoint.test.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+) => {
+  const result = compDict.closestPoint(context, [shapeType, shape], pt);
+  const g = secondaryGraph(result.contents);
+  const f = genCode(g);
+  const [x, y] = f([]).secondary;
+  expect(x).toBeCloseTo(expected[0]);
+  expect(y).toBeCloseTo(expected[1]);
+};
+
+export const testCircle = (
```
Comment (cmumatt) [inline packages/core/src/contrib/ClosestPoint.test.ts:27]: May be helpful to add a brief /** */ comment to each function to document its meaning and the meaning of its input / output values.

File: packages/core/src/contrib/ClosestPoint.test.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  const shape = makeCircle(context, canvas, {
+    center: vectorV(center),
+    r: floatV(radius),
+    strokeWidth: floatV(strokeWidth),
+    strokeColor: black(),
+  });
+  compareClosestPoint(context, "Circle", shape, pt, expected);
+};
+
+export const testRectangle = (
```
Comment (cmumatt) [inline packages/core/src/contrib/ClosestPoint.test.ts:44]: May be helpful to add a brief /** */ comment to each function to document the function and the meaning of its input / output values.

File: packages/core/src/contrib/ClosestPoint.test.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    center: vectorV(center),
+    width: floatV(width),
+    height: floatV(height),
+    strokeWidth: floatV(strokeWidth),
+    strokeColor: black(),
+  });
+  compareClosestPoint(context, "Rectangle", shape, pt, expected);
+};
+
+describe("closest point", () => {
```
Comment (cmumatt) [inline packages/core/src/contrib/ClosestPoint.test.ts:63]: May be helpful to add a /** */ comment here to briefly explain what the test is doing.  No need to get super detailed; a little will go a long way.

File: packages/core/src/contrib/Functions.ts [hunk 6]
```diff
@@ -1550,6 +1549,106 @@ export const compDict = {
...
+    /**
+     * Helper function for Rectangles
+     * @param _context
+     * @param x
+     * @param a
+     * @param b
+     * @returns
+     */
+    const clamp = (
+      _context: Context,
```
Comment (cmumatt) [inline packages/core/src/contrib/Functions.ts:1583]:
```suggestion
```
If not needed, may want to remove this parameter.

File: packages/core/src/contrib/Functions.ts [hunk 7]
```diff
@@ -1550,6 +1549,106 @@ export const compDict = {
...
+      x: ad.Num,
+      l: ad.Num,
+      u: ad.Num
+    ): ad.Num => {
+      return max(l, min(u, x));
+    };
+    const getNearestPointRect = (
```
Comment (cmumatt) [inline packages/core/src/contrib/Functions.ts:1590]: May be helpful to add a brief /** */ comment to each function to document the function and the meaning of its input / output values.

File: packages/core/src/contrib/Functions.ts [hunk 8]
```diff
@@ -1550,6 +1549,106 @@ export const compDict = {
...
+      w: ad.Num,
+      h: ad.Num,
+      x: ad.Num,
+      y: ad.Num
+    ): ad.Num[] => {
+      const r = add(l, w);
+      const b = add(t, h); //Formatted using JavaGraphics coordinate system
+      console.log(r);
+      console.log(b);
+      x = clamp(_context, x, l, r);
```
Comment (cmumatt) [inline packages/core/src/contrib/Functions.ts:1603]:
```suggestion
      x = clamp(x, l, r);
```
Can remove context argument if no longer needed in clamp's function body.

File: packages/core/src/contrib/Functions.ts [hunk 9]
```diff
@@ -1631,6 +1730,7 @@ export const sdEllipseAsNums = (
     abUnswizzled[1]
   );
   // float l = ab.y*ab.y - ab.x*ab.x;
+
```
Comment (cmumatt) [inline packages/core/src/contrib/Functions.ts:1733]:
```suggestion
```
Looks like an extra line was added here.  If not needed, can just remove it.

**Review by cmumatt — APPROVED** [review]
This looks good!  Suggested some minor improvements, mostly comment-type items.  CI wants you to run yarn format in the monorepo root.  Still, great job on this PR!!

**Comment by samestep** [comment]
@cmumatt @joshsunshine I cleaned up this PR; do you think it's ready to merge? I haven't yet actually read through all the implementation code for calculating closest points to verify that they're correct.

**Comment by keenancrane** [comment]
One thing I notice is that `closestPointPolyline` (in `Functions.ts`) is bitwise identical to `closestPointLine`, i.e., the former really just computes the closest point on a single line segment, not a polyline.  It's otherwise invoked in the right places.  So, I'm just gonna kill the latter and stick to `closestPointLine` (which makes the code more understandable and less redundant).

**Comment by keenancrane** [comment]
Also updating file extensions

**Pull request merged**